src/loan/terms/simple/loan/pwn_simple_loan.cairo

```diff
@@ -81,100 +146,704 @@ mod PwnSimpleLoan {
+            lender_spec: LenderSpec
+        ) {
+            let loan = self.loans.read(refinancing_loan_id);
+            let erc721_dispatcher = ERC721ABIDispatcher {
+                contract_address: self.loan_token.read().contract_address
+            };
+            // @note: owner_of needs u256 and here the id is in felt252
+            let loan_owner = erc721_dispatcher.owner_of(refinancing_loan_id.try_into().unwrap());
+            let repayment_amount = self.get_loan_repayment_amount(refinancing_loan_id);
+            let (fee_amount, new_loan_amount) = fee_calculator::calculate_fee_amount(
+                self.config.read().get_fee(), loan_terms.credit.amount
+            );
+            let fee_amount: u256 = fee_amount.into();
+            let common = if (repayment_amount > new_loan_amount) {
+                new_loan_amount
+            } else {
+                repayment_amount
+            };
+
+            let surplus = if (new_loan_amount > repayment_amount) {
+                new_loan_amount - repayment_amount
+            } else {
+                0
+            };
+
+            let shortage = if (surplus > 0) {
+                0
+            } else {
+                repayment_amount - new_loan_amount
+            };
+
+            let should_transfer_common = loan_terms.lender != loan_owner
+                || (loan.original_lender == loan_owner
+                    && loan.original_source_of_funds != lender_spec.source_of_funds);
+
+            let mut credit_helper = loan_terms.credit;
+
+            if (lender_spec.source_of_funds != loan_terms.lender) {
+                credit_helper.amount = fee_amount + surplus;
+                if (should_transfer_common) {
+                    credit_helper.amount += common;
+                }
+                self._withdraw_credit_from_pool(credit_helper, loan_terms, lender_spec);
+            }
+
+            if (fee_amount > 0) {
+                credit_helper.amount = fee_amount;
+                self
+                    .vault
+                    ._push_from(
+                        credit_helper, loan_terms.lender, self.config.read().get_fee_collector()
+                    );
+            }
+
+            if (should_transfer_common) {
+                credit_helper.amount = common;
+                self.vault._pull(credit_helper, loan_terms.lender);
+            }
+
+            if (surplus > 0) {
+                credit_helper.amount = surplus;
+                self.vault._push_from(credit_helper, loan_terms.lender, loan_terms.borrower);
+            } else if (shortage > 0) {
+                credit_helper.amount = shortage;
+                self.vault._pull(credit_helper, loan_terms.borrower);
+            }
+            let credit_amount = shortage + if (should_transfer_common) {
+                common
+            } else {
+                0
+            };
+            // @note: in solidity this was in try catch? shall we have it in match or something?
```

Comment:
We can't use match since the tx will revert in any case if the transfer is not succesful. Leave like it is now